platforms/core-execution/snapshots/src/main/java/org/gradle/internal/vfs/impl/DefaultFileSystemAccess.java

```diff
@@ -129,77 +119,89 @@ public Optional<HashCode> readRegularFileContentHash(String location) {
+                    .orElseGet(readFromDisk)
+            ));
+    }
+
+    /**
+     * Takes a snapshot of the given location and filters it according to the given filter.
+     */
+    private Optional<FileSystemLocationSnapshot> snapshot(String location, SnapshottingFilter filter) {
+        ImmutableMap<String, FileSystemLocationSnapshot> alreadyStoredSnapshots = virtualFileSystem
             .findRootSnapshotsUnder(location)
             .collect(ImmutableMap.toImmutableMap(
                 FileSystemLocationSnapshot::getAbsolutePath,
                 Function.identity()
             ));
-        FileSystemLocationSnapshot knownExactSnapshot = previouslyKnownSnapshots.get(location);
-        if (knownExactSnapshot != null) {
-            return knownExactSnapshot;
+        FileSystemLocationSnapshot unfilteredSnapshot = alreadyStoredSnapshots.get(location);
+        if (unfilteredSnapshot != null) {
+            return filterSnapshot(filter, unfilteredSnapshot);
+        } else {
+            return snapshotAndReuse(location, filter, alreadyStoredSnapshots);
         }
-        return virtualFileSystem.store(location, vfsStorer -> {
+    }
+
+    /**
+     * Takes a snapshot of the given location and filters it according to the given filter, reusing previously known snapshots.
+     */
+    private Optional<FileSystemLocationSnapshot> snapshotAndReuse(String location, SnapshottingFilter filter, ImmutableMap<String, FileSystemLocationSnapshot> previouslyKnownSnapshots) {
+        return virtualFileSystem.storeWithAction(location, vfsStorer -> {
             File file = new File(location);
             FileMetadata fileMetadata = this.stat.stat(file);
+            FileSystemLocationSnapshot unfilteredSnapshot;
             switch (fileMetadata.getType()) {
                 case RegularFile:
                     HashCode hash = hasher.hash(file, fileMetadata.getLength(), fileMetadata.getLastModified());
-                    return vfsStorer.store(new RegularFileSnapshot(location, file.getName(), hash, fileMetadata));
+                    unfilteredSnapshot = vfsStorer.store(new RegularFileSnapshot(location, file.getName(), hash, fileMetadata));
+                    break;
                 case Missing:
-                    return vfsStorer.store(new MissingFileSnapshot(location, fileMetadata.getAccessType()));
+                    unfilteredSnapshot = vfsStorer.store(new MissingFileSnapshot(location, fileMetadata.getAccessType()));
+                    break;
                 case Directory:
-                    // This will only store the captured snapshot in the VFS if the filter did not match anything
-                    return directorySnapshotter.snapshot(
+                    // This will capture a filtered snapshot, and only store the captured snapshot in the VFS
+                    // if the filter did not filter out anything.
+                    return Optional.of(directorySnapshotter.snapshot(
```

Comment:
🤔 Would it also make sense that `directorySnapshotter.snapshot()` would return `unfilteredSnapshot` and we would filter it inside `snapshotAndReuse`?

Reply:
There is a use case where this wouldn't work: if you had wanted to snapshot a large directory, but only capture a some part of it. With the current solution you never have to snapshot the unwanted files.